File: pimlico/term/terms/sequence/sequence.cpp
#include "sequence.hpp"

#include "../choice/choice.hpp"

Sequence::Sequence() : Term(Term::Type::SEQUENCE) {}

Term *Sequence::parse(ParseBuffer &buffer, ErrorBuffer &errors) {
    Sequence *sequence = new Sequence();
    sequence->position = buffer.position;

    while(true) {

        Term *value = Choice::parse(buffer, errors);
        if(value == nullptr)
            return nullptr;
        sequence->values.push_back(value);

        buffer.skip(true);
        if(buffer.finished() || buffer.peek('\n') || buffer.peek(')'))
            break;
    }

    if(sequence->values.size() == 1)
        return sequence->values.back();

    return sequence;
}

File: source/pimlico/grammar/terms/sequence.py
from __future__ import annotations

from perivale import Buffer, ParseError

from pimlico.grammar import Term

from .choice import Choice


class Sequence(Term):

    def __init__(self, terms: list):
        self.terms = terms
    
    def __str__(self) -> str:
        return self.serialize()
    
    def serialize(self) -> str:
        terms = [term.serialize() for term in self.terms]
        return " ".join(terms)

    @staticmethod
    def parse(buffer: Buffer, errors: list) -> Sequence | None:
        
        term_position = buffer.copy_position()
        
        enclosed = False
        if buffer.match("(", consume=True):
            enclosed = True
            buffer.skip_space()

        terms = []
        while True:
        
            if not Term.present(buffer):
                raise Exception("expected a term")
            
            term = Choice.parse(buffer, errors)
            if term is None:
                return None
            terms.append(term)

            buffer.skip_space()
            if buffer.finished() or buffer.read() in "\n{)":
                break
        
        if enclosed and not buffer.match(")", consume=True):
            error = ParseError()
            error.add_excerpt("expected ')'",
                    buffer.excerpt())
            error.add_excerpt("to match this",
                    buffer.excerpt(term_position))
            errors.append(error)
            return None
        
        if len(terms) == 1:
            return terms[0]

        return Sequence(terms)
            

